README.md
# VRFBackupTool.py #
----------

## About ##
**VRFBackupTool.py** is a Python application that allows a user to back up the
(VRF) VPN tunnel configuration on a Cisco router.

## Dependencies ##
Exscript module [https://github.com/knipknap/exscript/](https://github.com/knipknap/exscript/)

## Requirements ##
1. This application is hard-coded to use the SSH2 protocol; If SSH v2 is not
   enabled on your router(s), you will need to:
   * Add `ip ssh version 2` to your Cisco router(s) configuration and any 
   associated access-list changes.
   or
   * Alter the `default_protocol` variable in the `get_hosts_from_file` function
   to use a different protocol enabled on your router(s).
2. A valid username/password.
3. Presence of a `settings.cfg` file containing any pre-configured settings
   specified by the user.  Should this file not exist, the application will
   create it using default values.
3. Presence of a `routers.txt` file containing a list, one per line, of 
   Hostnames or IP Addresses of routers on which VRF tunnels are configured.
   (If one is not provided, the application will create an example file for
   the user.)
4. Presence of a `index.txt` file. The application will create this file for
   it's own use, if it does not already exist.

## Assumptions ##
1. This application was written for use on Cisco IOS devices and cannot be
   guaranteed to work on other makes/model routers.
2. This application assumes that you have enable privileges on each router
   in order to execute the `show running-config` command.  If you do not
   have sufficient user privileges, this application will not work as
   designed.

## Limitations ##
1. This application uses the same username/password to access ALL routers. If
   your routers use unique usernames/passwords, then this script will not work.

## Functionality ##
![](VRFBackupTool.png)

1. Upon execution, the application will search its parent directory for the
   presence of a file named `settings.cfg`.
   * If this file does not exist, the application will continue to Step 2.
   * If this file does exist, the application will proceed to Step 3.
2. The application will create a `settings.cfg` file with default values
   and continue to Step 3.
3. The application will open the `settings.cfg` file and read its default
   values into variables and continue to Step 4.
4. The application will search its parent directory for the presence of a file
   defined by the `routerFile` variable in `settings.cfg`.
   * If this file does exist, the application will continue to Step 5.
   * If this file does not exist, the application will proceed to Step 13.
5. The application will search its parent directory for the presence of a file
   defined by the `indexFile` variable in `settings.cfg`.
   * If this file does exist, the application will continue to Step 6.
   * If this file does not exist, the application will proceed to Step 10.
6. The application will check the timestamp of the last modification made to
   the indexFile.
   * If the file was created today, the application will continue to Step 7.
   * If the file was not created today, the application will proceed to Step 9.
7. The application will ask the user to provide the search criteria; Either
   the VRF name or Peer IP Address of the VPN tunnel to locate.  The application 
   will open the indexFile as read-only and search through the file.
   * If the search criteria was not found, exit program (END)
   * If the search criteria was found, continue to Step 8.
8. The application will ask the user for their username and password, should
   these values not be pre-configured in the `settings.cfg` file and log into
   each router containing the search string.  Using the
   `show running-config | section <VRFName>` and 
   `show running-config | section SMVPN <RouerDistinguisher>` commands, the
   application will backup the VRF configuration to a file in a directory
   specified in `settings.cfg`.  Should this filename already exist, the
   application will continue to append an integer to the end of the filename
   until it is no longer attmepting to overwrite an existing file.
9. The application will ask the user if they would like to update the indexFile.
   * If the user does not want to update the indexFile, the application will 
   proceed to Step 7.
   * If the user wants to update the indexFile, the application will continue 
   to Step 10.
10. The application will ask the user for their username and password, should
    these values not be pre-configured in the `settings.cfg` file.  These
    credentials will be used to login to each of the routers listed in the 
    routerFile.  Continue to Step 11.
11. The application will log in to each of the routers listed in the routerFile
    using the username and password provided by the user, should these values
	not be pre-configured in the `settings.cfg` file. Using the 
	`show run | include crypto keyring` command, the application will collect 
	all the VRFs configured on each router and write them to indexFileTmp.  
	Continue to Step 12
12. Removing all the unnecessary information from indexFileTmp, the application
    will copy the VRF Name, Customer Peer IP and Local Peer IP (separated by 
	commas) into indexFile.  The application will then remove the temporary file,
	indexFileTmp.  Return to Step 5.
    Note: If indexFile already exists, it will be overwritten during this step.
13. The application will display instructions for the use of the routerFile.  
	The application will also create an example file and place it in its parent
	directory end exit (END).


CHANGELOG.md
# VRFBackupTool.py #
---

## VRFBackupTool v0.0.10-alpha (2015-02-16) ##
* Made a minor correction to code to ensure compatability with 15.X train of
  Cisco IOS.

## VRFBackupTool v0.0.9-alpha (2014-03-17) ##
* Replaced tab with four spaces.
* Replaced ' with " to be consistent throughout the file.
* Corrected problem where application would fail if logFileDirectory or 
  backupFileDirectory in settings.cfg was blank.

## VRFBackupTool v0.0.8-alpha (2013-09-09) ##
* Corrected makedirs() functionality: Directories with a trailing backslash
  in the config file were not being created thereby causing the application
  to fail.
* Moved logFileDirectory & backupDirectory makedirs() function such that the
  directory would only be created if/when the parent function was called
  instead of creating both directories whenever the application executed.
* Removed 'dated' variable, which was a duplicate of global variable 'date'
 
## VRFBackupTool v0.0.7-alpha (2013-08-29) ##
* Updated the backupVRF function so that the application will not log into
  any routers and retrieve results if the output file cannot be opened.
* Corrected 'mkdir' function to 'makedirs' so that directories will be
  created recursively, if they do not exist.
* Added basic logging to file to track results if application has to connect
  to a router to run buildIndex().
* Suppressed error SPAM from stdout by adding stderr=(open(os.devnull, 'w'))
  to the Queue() function. (Errors are still written to the log.)

## VRFBackupTool v0.0.6-alpha (2013-08-28) ##
* Added functionality to specify configFile from the command line.
* Updated README.md

## VRFBackupTool v0.0.5-alpha (2013-08-26) ##
* Created README.md, VRFBackupTool.png
* Updated code to remove unused modules and add additional comments.
 
 ## VRFBackupTool v0.0.4-alpha (2013-08-22) ##
* Added whitespace to the end of the "show running-config | section SMVPN"
  command to prevent incorrect matches.  (Without the space at the end, a
  search for "3" would also return "31" or "300", etc.)

## VRFBackupTool v0.0.3-alpha (2013-08-20) ##
* Adjusted output spacing (removing/moving 'print' statements)
* Added additional comments to code, configFile, routerFile.

## VRFBackupTool v0.0.2-alpha (2013-08-16) ##
* Corrected bad REGEX being applied to routeDistinguisher that incorrectly
  caused the function to grab ALL SMVPN profiles instead of the only matching
  profile.
* Changed the configFile format in such a way that a single configFile
  can be used for my different VRF applications.
* Replaced old file open function method with new file open technique
  outputFile=open(file, 'w') --> with open(file, 'w') as openFile:
* Added error checking to append a trailing backslash to the backupDirectory
  should it be missing.
* Added error checking to append a trailing integer to the backup outputFileName
  should the file already exist.
* Updated error messages so they reflect the actual filename as read
  from the configFile.
* Rewrote "Building index..." message so it does not take up most of the
  screen when working with a large batch of routers.
* Updated example configFile output to reflect new changes in configFile format.

## VRFBackupTool v0.0.1-alpha (2013-08-15) ##
* Initial commit


VRFBackupTool.py
#!/usr/bin/env python
#
# VRFBackupTool.py
# Copyright (C) 2013-2015 Aaron Melton <aaron(at)aaronmelton(dot)com>
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


import argparse     # Required to read arguments from the command line
import base64       # Required to decode password
import ConfigParser # Required for configuration file
import datetime     # Required for date format
import Exscript     # Required for SSH, queue functionality
import re           # Required for REGEX operations
import sys          # Required for printing without newline
import os           # Required to determine OS of host

from argparse                   import ArgumentParser, RawDescriptionHelpFormatter
from base64                     import b64decode
from ConfigParser               import ConfigParser
from datetime                   import datetime
from Exscript                   import Account, Queue, Host, Logger
from Exscript.protocols         import SSH2
from Exscript.util.file         import get_hosts_from_file
from Exscript.util.log          import log_to
from Exscript.util.decorator    import autologin
from Exscript.util.interact     import read_login
from Exscript.util.report       import status,summarize
from re                         import search, sub
from sys                        import stdout
from os                         import getcwd, makedirs, name, path, remove, system


class Application:
# This class was created to provide me with an easy way to update application
# details across all my applications.  Also used to display information when
# application is executed with "--help" argument.
    author = "Aaron Melton <aaron@aaronmelton.com>"
    date = "(2015-02-16)"
    description = "Backs up the (VRF) VPN tunnel configuration on a Cisco router."
    name = "VRFBackupTool.py"
    url = "https://github.com/aaronmelton/VRFBackupTool"
    version = "v0.0.10-alpha"


def backupVRF(vrfName, localPeer):
# This function takes the VRF Name and Local Peer IP as determined during
# the searchIndex() function, retrieves all matching VRFs from their respective
# routers and writes the config to a file.

    # If backupDirectory does not exist, create it
    if not path.exists(backupDirectory): makedirs(backupDirectory)

    # Define output filename based on hostname and date
    outputFilename = backupDirectory+vrfName+"_Config_"+date+".txt"
    
    # Check to see if outputFilename currently exists.  If it does, append an
    # integer onto the end of the filename until outputFilename no longer exists
    incrementFilename = 1
    while fileExist(outputFilename):
        outputFilename = backupDirectory+vrfName+"_Config_"+date+"_"+str(incrementFilename)+".txt"
        incrementFilename = incrementFilename + 1
    
    with open(outputFilename, "w") as outputFile:
        try:
            if username == "":          # If username is blank
                print
                account = read_login()  # Prompt the user for login credentials
            
            elif password == "":        # If password is blank
                print
                account = read_login()  # Prompt the user for login credentials
        
            else:                       # Else use username/password from configFile
                # base64 decode password from the config file
                account = Account(name=username, password=b64decode(password))
                
            print
            print "--> Logging into "+localPeer+"..."
            
            socket = SSH2()             # Set connection type to SSH2
            socket.connect(localPeer)   # Open connection to router
            socket.login(account)       # Authenticate on the remote host
            
            print "--> Backing up "+vrfName+"..."
            
            socket.execute("terminal length 0") # Disable page breaks in router output
                                                # socket.autoinit() doesn't seem to disable
                                                # page breaks; Using standard command instead
            # Send command to router to retrieve first part of VRF configuration
            socket.execute("show running-config | section "+vrfName)

            outputFile.write(socket.response)   # Write contents of running config to output file
            
            # Use REGEX to locate Route Distinguisher in results from router
            routeDistinguisher = search(r"\srd\s\b[0-9]{0,4}\b:0", socket.response).group(0)
            # Use REGEX to remove everything but the actual Route Distinguisher number.
            routeDistinguisher = sub(r"\srd\s", "", routeDistinguisher)
            routeDistinguisher = sub(r":0", "", routeDistinguisher)
            
            # Send command to router to retrieve second part of VRF configuration
            socket.execute("show running-config | section SMVPN "+routeDistinguisher+" ")
            outputFile.write(socket.response)   # Write contents of running config to output file
        
            socket.send("exit\r")   # Send the "exit" command to log out of router gracefully
            socket.close()          # Close SSH connection

        # Exception: outputFile file could not be opened
        except IOError:
            print
            print "--> An error occurred opening "+outputFile+"."    

    print "--> "+vrfName+" backed up to "+outputFilename+"."

logger = Logger()   # Log stuff
@log_to(logger)     # Logging decorator; Must precede buildIndex!
@autologin()        # Exscript login decorator; Must precede buildIndex!
def buildIndex(job, host, socket):
# This function builds the index file by connecting to the router and extracting all
# matching sections.  I chose to search for "crypto keyring" because it is the only
# portion of a VPN config that contains the VRF name AND Peer IP.  Caveat is that
# the program temporarily captures the pre-shared key.  "crypto isakmp profile" was not
# a suitable query due to the possibility of multiple "match identity address" statements

    stdout.write(".")                   # Write period without trailing newline
    socket.execute("terminal length 0") # Disable user-prompt to page through config
                                        # Exscript doesn't always recognize Cisco IOS
                                        # for socket.autoinit() to work correctly

    # Send command to router to capture results
    socket.execute("show running-config | section crypto keyring")

    with open(indexFileTmp, "a") as outputFile:
        try:
            outputFile.write(socket.response)   # Write contents of running config to output file
        except IOError:
            print
            print "--> An error occurred opening "+indexFileTmp+"."

    socket.send("exit\r")   # Send the "exit" command to log out of router gracefully
    socket.close()          # Close SSH connection

    cleanIndex(indexFileTmp, host)  # Execute function to clean-up the index file
    
def cleanIndex(indexFileTmp, host):
# This function strips all the unnecessary information collected from the router leaving
# only the VRF name, remote Peer IP and local hostname or IP

    try:
        # If the temporary index file can be opened, proceed with clean-up
        with open(indexFileTmp, "r") as srcIndex:

            try:
                # If the actual index file can be opened, proceed with clean-up
                # Remove unnecessary details from the captured config
                with open(indexFile, "a") as dstIndex:
                    # Use REGEX to step through config and remove everything but
                    # the VRF Name, Peer IP & append router hostname/IP to the end
                    a = srcIndex.read()
                    b = sub(r"show running-config \| section crypto keyring.*", "", a)
                    c = sub(r"crypto keyring ", "", b)
                    d = sub(r".(\r?\n)..pre-shared-key.address.", ",", c)
                    e = sub(r".key.*\r", ","+host.get_name(), d)
                    f = sub(r".*#", "", e)
                    dstIndex.write(f)

            # Exception: actual index file could not be opened
            except IOError:
                print
                print "--> An error occurred opening "+indexFile+"."

    # Exception: temporary index file could not be opened
    except IOError:
        print
        print "--> An error occurred opening "+indexFileTmp+"."
    
    # Always remove the temporary index file
    finally:
        remove(indexFileTmp)    # Critical to remove temporary file as it contains passwords!

def confirm(prompt="", defaultAnswer="y"):
# This function prompts the user to answer "y" for yes or "n" for no
# Returns true if the user answers Yes, false if the answer is No
# The user will not be able to bypass this function without entering valid input: y/n

    while True:
        # Convert response to lower case for comparison
        response = raw_input(prompt).lower()
    
        # If no answer provided, assume Yes
        if response == "":
            return defaultAnswer
    
        # If response is Yes, return true
        elif response == "y":
            return True
    
        # If response is No, return false
        elif response == "n":
            return False
    
        # If response is neither Yes or No, repeat the question
        else:
            print "Please enter y or n."

def fileExist(fileName):
# This function checks the parent directory for the presence of a file
# Returns true if found, false if not

    try:
        # If file can be opened, it must exist
        with open(fileName, "r") as openedFile:
            return True # File found

    # Exception: file cannot be opened, must not exist
    except IOError:
        return False    # File NOT found

def routerLogin():
# This function prompts the user to provide their login credentials and logs into each
# of the routers before calling the buildIndex function to extract relevant portions of
# the router config.  As designed, this function actually has the capability to login to
# multiple routers simultaneously.  I chose to not allow it to multi-thread given possibility
# of undesirable results from multiple threads writing to the same index file simultaneously

    try:# Check for existence of routerFile; If exists, continue with program
        with open(routerFile, "r"): pass
        
        # Read hosts from specified file & remove duplicate entries, set protocol to SSH2
        hosts = get_hosts_from_file(routerFile,default_protocol="ssh2",remove_duplicates=True)

        if username == "":          # If username is blank
            print
            account = read_login()  # Prompt the user for login credentials

        elif password == "":        # If password is blank
            print
            account = read_login()  # Prompt the user for login credentials

        else:                       # Else use username/password from configFile
            account = Account(name=username, password=b64decode(password))
        
        # Minimal message from queue, 1 threads, redirect errors to null
        queue = Queue(verbose=0, max_threads=1, stderr=(open(os.devnull, "w")))
        queue.add_account(account)              # Use supplied user credentials
        print
        stdout.write("--> Building index...")   # Print without trailing newline
        queue.run(hosts, buildIndex)            # Create queue using provided hosts
        queue.shutdown()                        # End all running threads and close queue
        
        # If logFileDirectory does not exist, create it.
        if not path.exists(logFileDirectory): makedirs(logFileDirectory)

        # Define log filename based on date
        logFilename = logFileDirectory+"VRFBackupTool_"+date+".log"
        
        # Check to see if logFilename currently exists.  If it does, append an
        # integer onto the end of the filename until logFilename no longer exists
        incrementLogFilename = 1
        while fileExist(logFilename):
            logFilename = logFileDirectory+"VRFBackupTool_"+date+"_"+str(incrementLogFilename)+".log"
            incrementLogFilename = incrementLogFilename + 1

        # Write log results to logFile
        with open(logFilename, "w") as outputLogFile:
            try:
                outputLogFile.write(summarize(logger))

            # Exception: router file was not able to be opened
            except IOError:
                print
                print "--> An error occurred opening "+logFileDirectory+logFile+"."

    # Exception: router file could not be opened
    except IOError:
        print
        print "--> An error occurred opening "+routerFile+"."

def searchIndex(fileName):
# This function searches the index for search string provided by user and
# returns the results, if any are found

    # Ask the user to provide search string
    print
    searchString = raw_input("Enter the VRF Name you want to back up: ")
    
    # Repeat the question until user provides ANY input
    while searchString == "":
        searchString = raw_input("Enter the VRF Name you want to back up: ")
    
    # As long as the user provides ANY input, the application will search for it
    else:
        try:
            # If the index file can be opened, proceed with the search
            with open(fileName, "r") as openedFile:
                # Quickly search the file for search string provided by user
                # If search string found in the file, we will search again to return the results
                # Otherwise inform the user their search returned no results
                if searchString in openedFile.read():
                    openedFile.seek(0)    # Reset file cursor position
                    searchFile = openedFile.readlines()    # Read each line in the file one at a time

                    for line in searchFile:
                        if searchString in line:
                            word = line.split(",")          # Split up matching line at the comments
                            vrfName = word[0].strip()       # Strip out VRF name from search results
                            localPeer = word[2].rstrip()    # Strip out Local Peer IP from search results
                            backupVRF(vrfName, localPeer)
    
                # Else: Search string was not found
                else:
                    print
                    print "--> Your search string was not found in "+indexFile+"."
    
        # Exception: index file could not be opened
        except IOError:
            print
            print "--> An error occurred opening "+indexFile+"."
                             
def upToDate(fileName):
# This function checks the modify date of the index file
# Returns true if file was last modified today, false if the file is older than today

    # If the modify date of the file is equal to today's date
    if datetime.fromtimestamp(path.getmtime(fileName)).date() == datetime.today().date():
        return True    # File is "up to date" (modified today)

    # Else the modify date of the index file is not today's date
    else:
        return False    # File is older than today's date


# Check to determine if any arguments may have been presented at the command
# line and generate help message for "--help" switch
parser = ArgumentParser(
    formatter_class=RawDescriptionHelpFormatter,description=(
        Application.name+" "+Application.version+" "+Application.date+"\n"+
        "--\n"+
        "Description: "+Application.description+"\n\n"+
        "Author: "+Application.author+"\n"+
        "URL:    "+Application.url
    ))
# Add additional argument to handle any optional configFile passed to application
parser.add_argument("-c", "--config", dest="configFile", help="config file", default="settings.cfg", required=False)
args = parser.parse_args()      # Set "args" = input from command line
configFile = args.configFile    # Set configFile = config file from command line OR "settings.cfg"

# Determine OS in use and clear screen of previous output
if name == "nt":    system("cls")
else:               system("clear")

# PRINT PROGRAM BANNER
print Application.name+" "+Application.version+" "+Application.date
print "-"*(len(Application.name+Application.version+Application.date)+2)

# START PROGRAM
# Steps below refer to documented program flow in VRFBackupTool.png
# Step 1: Check for presence of settings.cfg file
try:
# Try to open configFile
    with open(configFile, "r"):    pass
    
# Step 2: Create example settings.cfg file
except IOError:
# Except if configFile does not exist, create an example configFile to work from
    try:
        with open (configFile, "w") as exampleFile:
            print
            print "--> Config file not found; Creating "+configFile+"."
            print
            exampleFile.write("## VRFBackupTool.py CONFIGURATION FILE ##\n#\n")
            exampleFile.write("[account]\n#password is base64 encoded! Plain text passwords WILL NOT WORK!\n#Use website such as http://www.base64encode.org/ to encode your password\nusername=\npassword=\n#\n")
            exampleFile.write("[VRFBackupTool]\n#variable=C:\path\\to\\filename.ext\nrouterFile=routers.txt\nindexFile=index.txt\nindexFileTmp=index.txt.tmp\nlogFileDirectory=\nbackupDirectory=\n")

    # Exception: file could not be created
    except IOError:
        print "--> An error occurred creating the example "+configFile+"."

# Step 3: Open settings.cfg file and read options
finally:
# Finally, using the provided configFile (or example created), pull values
# from the config and login to the router(s)
    config = ConfigParser(allow_no_value=True)
    config.read(configFile)
    username = config.get("account", "username")
    password = config.get("account", "password")
    routerFile = config.get("VRFBackupTool", "routerFile")
    indexFile = config.get("VRFBackupTool", "indexFile")
    indexFileTmp = config.get("VRFBackupTool", "indexFileTmp")
    logFileDirectory = config.get("VRFBackupTool", "logFileDirectory")
    backupDirectory = config.get ("VRFBackupTool", "backupDirectory")
    
    # If logFileDirectory is blank, use current working directory
    if logFileDirectory == "":  logFileDirectory = getcwd()

    # If logFileDirectory does not contain trailing backslash, append one
    if logFileDirectory != "":
        if logFileDirectory[-1:] != "\\": logFileDirectory = logFileDirectory+"\\"
            
    # If backupDirectory is blank, use current working directory
    if backupDirectory == "":  backupDirectory = getcwd()

    # If backupDirectory does not contain trailing backslash, append one
    if backupDirectory != "":
        if backupDirectory[-1:] != "\\": backupDirectory = backupDirectory+"\\"
            
    # Define "date" variable for use in the output filename
    date = datetime.now()    # Determine today's date
    date = date.strftime("%Y%m%d")  # Format date as YYYYMMDD

    # Step 4: Check for presence of routerFile
    # Does routerFile exist?
    if fileExist(routerFile):
        # Step 5: Check for presence of indexFile file
        # Does indexFile exist?
        if fileExist(indexFile):
            # Step 6: Check date of file
            # File created today?
            if upToDate(indexFile):
                # Step 7: Prompt user to provide backup criteria
                # Step 8: Backup VRF configuration, if any and END PROGRAM
                print
                print("--> Index found and appears up to date.")
                searchIndex(indexFile)
            else: # if upToDate(indexFile):
                # Step 9: Ask user if they would like to update indexFile
                # Update indexFile?
                print
                if confirm("--> The index does not appear up to date.\n\nWould you like to update it? [Y/n] "):
                    # Step 10: Prompt user for username & password (if not stored in settings.cfg)
                    # Step 11: Login to routers and retrieve VRF, Peer information
                    # Step 12: Sort indexFile to remove unnecessary data
                    # GOTO Step 2 (Check for presence of indexFile file)
                    # Remove old indexFile to prevent duplicates from being added by appends
                    remove(indexFile)
                    routerLogin()
                    print
                    searchIndex(indexFile)
                else: # if confirm("Would you like to update the index? [Y/n] "):
                    # GOTO Step 7: (Prompt user to provide search string)
                    # Step 8: Backup VRF configuration, if any and END PROGRAM
                    searchIndex(indexFile)
        else: # if fileExist(indexFile):
            # Step 10: Prompt user for username & password (if not stored in settings.cfg)
            # Step 11: Login to routers and retrieve VRF, Peer information
            # Step 12: Sort indexFile to remove unnecessary data
            # GOTO Step 2 (Check for presence of indexFile file)
            print("--> No index file found, we will create one now.")
            routerLogin()
            print
            searchIndex(indexFile)
    else: # if fileExist(routerFile):
        # Step 13: Create example routerFile and exit program and END PROGRAM    
        try:
            with open (routerFile, "w") as exampleFile:
                print
                print "--> Router file not found; Creating "+routerFile+"."
                print "    Edit this file and restart the application."
                exampleFile.write("## VRFBackupTool.py ROUTER FILE ##\n#\n")
                exampleFile.write("#Enter a list of hostnames or IP Addresses, one per line.\n#For example:\n")
                exampleFile.write("192.168.1.1\n192.168.1.2\nRouterA\nRouterB\nRouterC\netc...")

        # Exception: file could not be created
        except IOError:
            print
            print "--> Required file "+routerFile+" not found; An error has occurred creating "+routerFile+"."
            print "This file must contain a list, one per line, of Hostnames or IP addresses the"
            print "application will then connect to download the running-config."

print
print "--> Done."
raw_input() # Pause for user input.